neo.UnitTests/Network/P2P/Payloads/UT_Transaction.cs

```diff
@@ -275,5 +275,42 @@ public void ToJson()
             jObj["script"].AsString().Should().Be("4220202020202020202020202020202020202020202020202020202020202020");
             jObj["sys_fee"].AsNumber().Should().Be(42);
         }
+
+        [TestMethod]
+        public void Infinite_Loop_Not_Allowed_On_Verification()
+        {
+            byte[] script;
+            using (ScriptBuilder sb = new ScriptBuilder())
+            {
+                sb.Emit(OpCode.JMP, new byte[] { 0, 0, 0, 0 });
+                script = sb.ToArray();
+            }
+
+            long netfee = 100000000; // network fee
+
+            // Check Application
+            long appGas = 0;
+            using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Application, null, null, netfee, false))
+            {
+                engine.LoadScript(script);
+                Assert.AreEqual(VMState.FAULT, engine.Execute());
+                Assert.AreEqual(0, engine.ResultStack.Count);
+                appGas += engine.GasConsumed;
+            }
+            // many gas are spent before fault
+            Assert.AreEqual(appGas, 100000040);
+
+            // Check Verification
+            long verificationGas = 0;
+            using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Verification, null, null, netfee, false))
+            {
+                engine.LoadScript(script);
+                Assert.AreEqual(VMState.FAULT, engine.Execute());
+                Assert.AreEqual(0, engine.ResultStack.Count);
+                verificationGas += engine.GasConsumed;
+            }
+            // no gas is spent (backwards jump is not allowed)
+            Assert.AreEqual(verificationGas, 0);
```

Comment:
@erikzhang @shargon this basic test is already giving evidence that, for Verification, zero gas is spent (no backward jumps allowed), but on Application, it goes until funds are exhausted (and fails. this wouldn't consume the funds on verification, and this could be repeated again and again).

Reply:
How can we prevent the issue with CALLS?

Reply:
On CALLS, we have two options... count StepInto(), or just count it on PreExecute(). Must see what's better.